### DISCUSSION TITLE
Something confuses me when I  repeat the tutorials from SSCHA school 2023

### DISCUSSION THREAD

**Discussion by jgSi1113** [SSCHAcode/python-sscha#198] [Q&A] [answered]
Dear SSCHA developers,

Recently, I started to study the SSCHA according to the turorials from SSCHA school 2023 by myself, but the following thing conduses me.

1. What dose the "Good minimization steps" means?
The first turorial is the free energy minimization of Gold.  I set the the number of configurations and iterations as following:
_**N_CONFIGS = 50
MAX_ITERATIONS = 20**_
When the calculation finished and I plotted the minimization data in Fig.1, I found the _Good minimization steps_ is almost 60. How can I obtain the value of 60?  And I found the maximum value from the turorial is about 50 (http://sscha.eu/Tutorials/tutorial_01_first_simulations/), what causes this difference?  In addtion, when I change the **_calculator_** into QE, the value is just larger than 10 (see in Fig.2).
![图片1](https://github.com/SSCHAcode/python-sscha/assets/50296707/56d770fc-631a-4afc-92c4-c3a345d9dc4f)
Fig.1  Minimization data of Gold from sscha.
![图片2](https://github.com/SSCHAcode/python-sscha/assets/50296707/8b2507eb-7132-486f-bc50-6c9997b03fd7)
Fig.2 Minimization data of Gold via QE.

2. What is the default _**convergence criteria**_ for a minimization?
When I repeat the free energy minimization for H3S, I set the _**number_of_configurations**_ firstly. When all calculations well finished, I found the log for free energy minimization says as Fig. 3. This seems to have converged, but **_there is no minim*.dat and minim*.freq created_**.  Why is this happening? The turorial say "_This means that the dynamical matrix changed a sufficient amount that the original ensemble was not good enough anymore to describe the free energy of the new dynamical matrix; therefore, a new ensemble should be extracted._" . Therefore, I re-grnerated the ensemble by  _**number_of_configurations = 100**_  for pop2 and pop3. Both pop2 and pop3 used the dyn files from the pervious step.  The log for pop2 and pop3 is shown in Figs. 4 and 5. It seems that pop2 does not converge and pop3 is converged. I lpotted all phonon dispersions as shown in Fig. 6. One can see that phonon dispersion for pop1 is obviously different from the pop3. So, why the pop1 and pop3 are both converged?  For pop3, the free energy not decrease monotonically，will it have an impact on the results? 
![图片3](https://github.com/SSCHAcode/python-sscha/assets/50296707/7f7ea82d-e3f3-449e-9795-21cf50911be6)
Fig.3 Log for free energy minimization calculating for pop1.
![图片4](https://github.com/SSCHAcode/python-sscha/assets/50296707/1d573ba4-21b8-412f-8103-3b77725ea014)
Fig.4 Log for free energy minimization calculating for pop2.
![图片5](https://github.com/SSCHAcode/python-sscha/assets/50296707/e68f7ed9-80ab-4eea-bf4d-3651b07dfd14)
Fig.5 Log for free energy minimization calculating for pop3.
![QQ截图20240315165008](https://github.com/SSCHAcode/python-sscha/assets/50296707/069bc084-c1be-463d-8c83-0f6bd6fe42f6)
Fig.6 Phonon dispersion for H3S.

Thank and Regards!
Jianguo

**Comment by jgSi1113** [comment]
By the way, will SSCHA schools be held in 2024?

**Reply by mesonepigreco** [reply]
We are still organizing it. We will probably hold an online webinar school this year and, again, an entire in-person school next year, but it needs to be confirmed.

**Reply by jgSi1113** [reply]
Looking forward to exciting school！

Regards！

**Comment by diegomartinez2** [comment]
Hi, first with the number of random configurations set so low (50) the total good minimization steps needed to reach the convergence could be also somewhat random.

Second, the SSCHA method is based on the minimization of the free energy (Gibbs-Bogoliubov variational principle). The code uses a preconditioned gradient descend method.

**Reply by diegomartinez2** [reply]
The criteria for convergence is taken from the gradients of the trial centroid positions and the trial harmonic matrix.

For more details on the convergence criteria you can look at J. Phys.: Condens. MAtter 33 (2021)

**Reply by jgSi1113** [reply]
Thank for your kind reply.

I will read it and try my best to understand it.

Cheers！
Jianguo

**Reply by diegomartinez2** [reply]
Also, the number of configurations in the examples are too low to give realistic results. You need more than 1000 configurations (and usually even bigger numbers), but that means more time expended in the calculations of the forces.

**Reply by jgSi1113** [reply]
Thank again.  If my understanding is correct, the more configurations, the better results we obtained.  And the number of configurations i used just depend on the computing resources i can pay for?  

Regards!

**Reply by diegomartinez2** [reply]
The number of configurations has a sweet spot. Increasing the number of configurations makes the results better but is not linear, after a certain point increasing the number of configurations does not add any meaningful improvement. Usually you take a number of configurations and test with a smaller or bigger one to see if it's worth it.

**Reply by jgSi1113** [reply]
Copy that!

Thank you so much!
Regards!
Jianguo

**Comment by mesonepigreco** [comment, marked as answer]
Thanks for your feedback.
1. The goal is to minimize the free energy. This is achieved by obtaining the gradients of the free energy as low as possible (lower than their stochastic error).  The number of steps necessary to converge is pretty random, as it depends on the step size (which is automatically optimized during the minimization) and the specific configuration extracted for each run. However, usually, steps are very fast so it is not that important. What is more important is the number of populations or recalculations of the energies and forces. This is set with MAX_ITERATIONS in your code and corresponds to the number of times the effective sample size restarts from 1 (6 times in your case). This is more important as it is the number of times the code recomputes the N_CONFIGS (50 in your case) configurations. So, the code evaluated 6 * 50 = 300 configurations of gold to converge. If those were done with quantum espresso for DFT, for example, it would be the overall computational cost of the calculation (as the sscha steps between different recalculations are speedy). 
2.  The convergence you can see from the values of the gradient. In POP1, you have ``FC gradient modulus = 81238.83 +- 59674020. `` So, for the code, it is lower than the stochastic error, and it says it is converged. However, it is very big (because the number of configurations used wasn't enough)! Indeed, when you compare it to the last one which is 48 +- 366 you see that the last one is converged to a much lower value. Which is good! The reason why you do not have the minim*.dat is because you should add the 
```
IO_freq = sscha.Utilities.IOInfo()
IO_freq.SetupSaving("minim")

# Initialize the SSCHA object relax [...]
relax.setup_custom_functions(custom_function_post = IO_freq.CFP_SaveAll)
```
Hope this helps!

**Reply by jgSi1113** [reply]
Thanks a lot.  I think I got the answer I wanted from your reply.

A minor thing is coming. The code I used in the steps of minimization as following,

```
import sscha, sscha.Ensemble, sscha.SchaMinimizer
import sscha.Utilities
import cellconstructor as CC, cellconstructor.Phonons

POPULATION = *

dyn = CC.Phonons.Phonons("dyn_end_population*_", 3)
ensemble = sscha.Ensemble.Ensemble(dyn, 0)
ensemble.load("data", population = POPULATION, N = 100)

minim = sscha.SchaMinimizer.SSCHA_Minimizer(ensemble)
minim.init()

\# Save the minimization details
**ioinfo = sscha.Utilities.IOInfo()
ioinfo.SetupSaving("minim_{}".format(POPULATION))

minim.run(custom_function_post = ioinfo.CFP_SaveAll)**
minim.finalize()
minim.dyn.save_qe("dyn_end_population{}_".format(POPULATION))
```

It work well for pop2 and pop3, but not works for pop1. Bug or something ignored? Or I must use the  _relax_?

Regareds!

**Reply by diegomartinez2** [reply]
For pop1 you start with the harmonic dynamical matrix, your code reads the dyn matrix with `dyn = CC.Phonons.Phonons("dyn_end_population*_", 3)` that in the case of pop1 it cannot read unless you rename the harmonic ones as `"dyn_end_population1_"`.

**Reply by mesonepigreco** [reply]
Ah, ok I see.
No, the simple problem is that in pop1, you have only 1 step of minimization (you can see that the minimization ends at the second step). For this reason, the code never finds a good step to save the frequencies and the minimization data. 
If something like this happens, try to use a much lower value of meaningful_factor:
```
minim.meaningful_factor = 0.0001
```
In this way, it should print the minimization data for the first population (and probably it will go out of the statistical ensemble).

``meaningful_factor`` sets the ratio between gradients and their error to consider converged a minimization.

**Reply by jgSi1113** [reply]
Cool, that's great for me. 

The minim*.dat is created by adding following parameters:
# Define the steps for the centroids and the force constants
_minim.min_step_dyn = 0.005         # The minimization step on the dynamical matrix
minim.min_step_struc = 0.05        # The minimization step on the structure
minim.kong_liu_ratio = 0.5          # The parameter that estimates whether the ensemble is still good
minim.meaningful_factor = 0.000001  # How much small the gradient should be before I stop_

Thanks for you kind discussion. 
Regards!

**Reply by mesonepigreco** [reply]
You are welcome! Happy to help
